### ISSUE TITLE
Printed types for some theorems are not reconstructible

### ISSUE THREAD

**Issue by Ailrun** [Beluga-lang/Beluga#207]
For example, when try to `session serialize` the following theorem,
```
Sn-subst : (g : cxt) {M : [g, x : tm A[] |- tm B[]]} Sn [g |- M[..,N]] -> Sn [g, x : tm A[] |- M]
```

Harpoon gives the next type which is not reconstructible.
```
proof Sn-subst : (g : cxt)
                   {M : (g, x : tm A[] |- tm B[])}
                     Sn [_] [ |- B] [_ |- M[.., N]] ->
                       Sn [_, x : tm A[]] [ |- B] [_, x1 |- M]
```

@tsani, could you tell me which one causes this problem? I mean, is this from the printer or from the reconstructor?

**Comment by tsani** [comment]
It's the pretty-printer. When printing a context variable, it checks
if that variable is implicit. If it is, it prints it as `_`. You've
shown that this analysis is essentially broken when
printing a comp type (and probably comp kinds also) because we have
this implicit context quantification `(g : ctx)`. This binds an
*implicit* variable, so the analysis thinks it should be printed as
`_`, but actually it needs to print it as `g`.

Here's how I think one can fix it.
When printing comp types (and kinds), when encountering a PiBox, cD
needs to be extended very carefully: implicit context quantification
needs to actually be considered *explicit*. This way, later
occurrences of `g` will actually print as `g` and not as `_`.

**Comment by Ailrun** [comment]
@tsani Thank you for confirming. I will check that part.

**Comment by tsani** [comment]
I came across the code that causes this issue in the pretty-printer today while working on something unrelated, so I figured I'd fix it.

Resolved by commit a4b469b4fd052b5d9ae3608ac35e97f7f4f917b3
